### PR TITLE
Fix: Clamped billboards did not render in 2D/CV scene modes

### LINKED ISSUE #5041: Ground clamped billboard in wrong location during scene transform
Reported on the forum: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/DjnveYPBjtw

https://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Ground%20Clamping.html&label=Showcases

When you transition to either 2D or CV, the billbaord and label are in the wrong location during the morph.  It sometimes snaps to the correct position after the morph completes, but sometimes it doesn't move until after you move the camera

### LINKED ISSUE #5042: Clamped billboard has incorrect bounding sphere in 2D and Columbus View
https://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Clamp%20to%20Terrain.html

- Switch to 2D or CV
- Double click the label to track it

The selection indicator is in the wrong place and the camera seems to be tracking wherever the selection indicator is pointing to

image

### LINKED ISSUE #8307: Billboard and Point have wrong location in 2D vs 3D
point

When switching from 2D to 3D, billboards and points that are clamped to ground will move to a different position. 

Possibly the same issue as https://github.com/AnalyticalGraphicsInc/cesium/issues/5041

Sandcastle example: https://sandcastle.cesium.com/#c=rVPRbtsgFP0V5Bc7koVTR31YmkbrnGp7aJupyfZkqbqBawcNQwQ4XTb134ft2Es6aZq08gSHe8/hHGAPhuwFPqMh10ThM8nQirqiX1ssClm7zLRyIBSacHSVq1wlCbnhnADZaaFcrjoGisoJJ9BS4Dz6mSvix05bj2lFpj11Bsb5GagJLYyuFlgaRBul45ik41Hct3niKTmStIj4jnIlfuCUXIzj3/gWRbl1j1igQcVwkPl0jtPs7ub+89N6+fTxcfnlYdERvOTq5ZWljZByo8Hwt7JFLxtj9LK3NgiQM3+ighI9FFKatHObFMCEFO5AS1GEJ5YtA9mUpp76rYKIhyQG3wwqNEALeVjr3jdH64SCxvq/XWhM3o2bMTrhX4HiDKyT2ES61lpuwHyondMqCieLMCZFrVijEY36iI5nsgwV0kpz9O/1qL9qsHsP0VV2+3A7WVwNWn9VSv9XKR2UgjiYWXeQOO843otqp40jtZGRv0+H1U6CDynZ1OwbOsqsbdpmSd8042JPBL/Og1c/Lg8Ik2Ct3ylq2X6APJjPEl9/1iY1cKHK5R6NhENTsr2Y33UgpXSW+OWfXa5L5ITxFw

Browser: Google Chrome 78.0.3904.17 (Official Build) dev (64-bit)

Operating System: Windows 10

### LINKED ISSUE #12531: Primitive with heightReference have a wrong positioning in 2D mode
### What happened?

Hello!

I use Cesium version 1.127.0.
When a primitive has a heightReference other than NONE, its position on map is wrong in 2D mode.
In 3D or without the heightReference, the primitive is placed properly.

If the primitive is displayed at lattitude longitude [0, 0], it is correctly placed.

Thank you for your help!

[Sandcastle with option to add a point with a heightReference set to CLAMP_TO_GROUND](https://sandcastle.cesium.com/#c=vZRtb9owEMe/ipVXqcQMG6voKEVDgEalUhCk48UyVca5EGuOjWyHlk189zlPlNBMqzRpSEj4fP/z7467o1Jog3YMnkChGyTgCQ1BsyTGXzOb6zs0Ow+lMIQJUL5zce0LX+QarCkIwLEMwMoL6TK1Ta0JL4fj+/GHUSag2VNbyYSZS80Mk+JFMiTK2F9EtHGoZDyCjQLQ7rvOJb781Ol0GuhjC7faV1ft/PUwETSLQIJgnoZ0L9AvXyC0JCKgRBsOOADKiQK3dMmVqMgWgzAWAjS2926mRZYuB+tWORvlrbV1UeGLEJVcqu4xhfSEV5Nbb9woPbbsGfiS/YQuusxth+zukLIcUpyaVFbMRBNgm8gsIAQFgsIbsqtR/ZeEF+NRbbpHY1SFOsrPYPHwbjCdP3qzxy+L2cP9qL5cJwWwWXhS8jVRUxCJ+y11LEgNPFts3xkEQZ4DWt16k9mDV8AgVb7qOwWnbU7gQK3s+H+URU8/L212XZpOUCIblKeBK712wt/4K9w/k9V2zdtg/9A6Vf7veTedBFKgwdgRJnov6Gu4ousoiUERHPL9RMbgtvLI5y2pIJY7GHDuVu7z7WJAKbt75kruWHC+p8acs62WLPCqXnWBNlyu08nZmsgDbQYb669NIbRhQ8I1pK1mv07D6Wmz59Ava/iZxVupDEoUdzFuGoi3nNit1Vwn9AcYTLUuy9Zrnkp7AdshFtzU7FJkh1hrexMmPJsc3+n3mtb/lZRLEjCxme1AcbJP3aL3/bvciDHuNe2xXmnyKTmL/Bs)

### Reproduction steps

1. Add two primitives (point or billboard) with and without a heightReference different than NONE
2. Set the map in 2D mode
3. The two primitives are not placed at the same position on map

### Sandcastle example

https://sandcastle.cesium.com/#c=vZRtb9owEMe/ipVXqcQMG6voKEVDgEalUhCk48UyVca5EGuOjWyHlk189zlPlNBMqzRpSEj4fP/z7467o1Jog3YMnkChGyTgCQ1BsyTGXzOb6zs0Ow+lMIQJUL5zce0LX+QarCkIwLEMwMoL6TK1Ta0JL4fj+/GHUSag2VNbyYSZS80Mk+JFMiTK2F9EtHGoZDyCjQLQ7rvOJb781Ol0GuhjC7faV1ft/PUwETSLQIJgnoZ0L9AvXyC0JCKgRBsOOADKiQK3dMmVqMgWgzAWAjS2926mRZYuB+tWORvlrbV1UeGLEJVcqu4xhfSEV5Nbb9woPbbsGfiS/YQuusxth+zukLIcUpyaVFbMRBNgm8gsIAQFgsIbsqtR/ZeEF+NRbbpHY1SFOsrPYPHwbjCdP3qzxy+L2cP9qL5cJwWwWXhS8jVRUxCJ+y11LEgNPFts3xkEQZ4DWt16k9mDV8AgVb7qOwWnbU7gQK3s+H+URU8/L212XZpOUCIblKeBK712wt/4K9w/k9V2zdtg/9A6Vf7veTedBFKgwdgRJnov6Gu4ousoiUERHPL9RMbgtvLI5y2pIJY7GHDuVu7z7WJAKbt75kruWHC+p8acs62WLPCqXnWBNlyu08nZmsgDbQYb669NIbRhQ8I1pK1mv07D6Wmz59Ava/iZxVupDEoUdzFuGoi3nNit1Vwn9AcYTLUuy9Zrnkp7AdshFtzU7FJkh1hrexMmPJsc3+n3mtb/lZRLEjCxme1AcbJP3aL3/bvciDHuNe2xXmnyKTmL/Bs

### Environment

Browser: Chrome and Edge
CesiumJS Version: 1.127.0
Operating System: Window 10 and 11

### PR DESCRIPTION
**Pull request by harrythemorris**
# Description

Terrain-clamped billboards were invisible in 2D and Columbus View (or, more accurately, drawn in the wrong place and failing depth test). Un-clamped billboards had peculiar rendering issues in 2D.

image

_Broken 2D left,  3D right_

And, lastly, the bounding sphere of clamped billboards in 2D/CV mode often covered the majority of the globe.
image

A few separate issues contributed to the problem

## Position / coordinate frame confusion

Initial debugging with spector.js revealed that no draw commands were submitted after the globe pass, which led me to suspect frustum culling as the culprit.

Logging the 'actualPosition` revealed that the position was outside of the expected range

| Cartographic | 3D | 2D |
| :--- | :--- | :--- |
| `(-2.1250727546554997, 0.8177268971321392, 4500)` | `(-2301506.550232819, -3718076.377732224, 4633834.269943927)` | `(-5232539.533463152, -8453150.669048188, 10574895.589471862)` |

In 2D and columbus, the coordinate is meant to be in the projected-map frame (swizzled so X=height), but what we got was _dramatically_ out of range height values.

There are two positions attached to billboards when clamped: 

The **clamped position** is the position on the terrain that the billboard should clamp to, and it is updated by `scene.updateHeight` callbacks.

The **actual position**, conversely, is meant to reflect the position that the billboard should actually render in the current rendering frame, as it is the vertex loaded onto the GPU and used for frustum culling.

The breakdown comes from the ambiguity of the coordinate frame - it seems like a lot of code was written assuming only ECEF exists in a bunch of places, which is not true and which I've now learnt debugging this issue 😁 

### Billboard Clamped Position
The space this is intended to be in is not clear as the `updateFunction` converted it to ECEF, but then acted upon it as if it were in the projected map frame as it incremented the X to apply a vertical offset.

`computeScreenSpacePosition` assumed that the coordinate was in the current mode frame, and did an unproject step to convert back to ECEF if not 3D. This was _not_ the case and the coordinate was already in ECEF, so did not require any transformation at all. This caused the screen position calculated to be very much incorrect too.

Fixes:
- The clamped position was changed to always be in ECEF, and the extra conversions removed.
- Actual position is kept as an accurate "rendered position". The naming was not changed to prevent any breakages, but could be improved to reduce ambiguity.

### Tile pick positions in update height callbacks
`QuadTreePrimitive.updateHeights` assumed coordinates coming back from `GlobeSurfaceTile.pick` were in ECEF. However they also reflect the current projected map frame. They were always passed into `cartesianToCartographic` to get a cartographic out, which converted assuming ECEF and gave the wrong cartographic back.

Fixes:
- Unproject/swizzle the position to get the cartographic instead of `cartesianToCartographic`.
- Update cache to store Cartographic to prevent caching of a different frame returning inappropriate positions.

## Three point check at smaller distances

When the camera was close to the billboard, it would not render, but draw commands were submitted. This meant it was likely getting discarded in the shader.

In 2D mode, the globe and billboard are (theoretically) coplanar. But in reality, `getGlobeDepth` is relatively imprecise when we are in 2D. The error it produced was _greater_ than the built in "depthsilon" of 10 units used in the three point check,  which caused the verts to be erroneously discarded.

This was confirmed with a debug shader colouring the billboards by this distance.

image

_3D, green in range, yellow within depthsilon_

image

_2D, all red = discarded_

`BillboardCollectionVS.glsl` has a max distance check when determining whether to use the three point check or not, which meant that this symptom did not show at moderate to far distances as the check wasn't run.

The proposed fix is to skip the three-point check when in 2D mode, as we are expecting all billboards to pass the depth check against the globe. (This may be a faulty assumption!)

## Issue number and link

Fixes #5042
Fixes #12531
Fixes #5041 (I believe, sandcastle and forum link are both dead)
Fixes #8307 

## Testing plan

Unit tests:
- `BillboardCollection` spec: a `CLAMP_TO_GROUND` billboard reports the **projected** map position (not the raw ECEF clamped position) in 2D.
- `QuadtreePrimitive` spec: the height cache stores a distinct **cartographic** per entry and is not mutated by callbacks.

Manually verified with [this sandcastle](https://sandcastle.cesium.com/#c=1VmNTxs7Ev9XprmTbvMubIBC2wPSHoTQhy4QBGmrU7fiObuTxMJr77O9gVzF/e0nf+xmNywfrd6ddEh8xJ4Zz/zm5/HYdLswRikJ5RsxI2mGCSwUcLH6OKGMTQSRiQISS6EUqBg5QioSVGHEu13zDX1G4xsgfHk7R4kgOOg5woyJCYIWkDESI+hbASmRNyjVntUC2ID+8PDsYnAMgcSkDeZrD+ZIZ3N9iVOUyGN0Mtfj0fXHy9Gn82Mg2pqPzaKYABO8y4gOS5vno/ON0m68JLz90Ob56HzQATGdKtTw761NSAGJ0h0vCIqkGcPE22z+mkqRgnYAghJANSiqVRm+FDlPQAsRRjziNM2E1PALEAV9VDRPnYGoFdtPUWvfiMWCKw0LircooQccb710+NmOBYV8X3BNKEcZtTrwPeJQuLJXKPjchmaZL0KyxA8E7U7E79v7xWIupT2/aGg/7kfc/g5tEsMEMz0fo9KHM0K50t4S9EDLHEtLGeVHOWXJuusX5XjQtlF2uwXod5lEpTABykGShBKuQExNUmdU5wlCIEU+m7Nlp8i749BfFDCirUg7LNYfnZxcDcbXZ4Px4PIKerC1GW5WUPWsPk3JzMS78tZC1BdMyMB7bD+El4PjDuy8a4daHBNNPl0OgxVsU3r3I7b6/zw8bzIWcYYa1FzcHhm+UD67yuwu6sGUMLUCt7IXey5nYSZpSjVdoApJkgSGBBXUjwqFvmAMY00FD777bN8bDqxiYWSC7FG7FZtDI9hoz8UyzbmdAJJlbHmMk3z2WbA8xaDtSLqKIkzM7FUlcCdpvHiAxn7E72v2JU4M3meuoAQ+tRdCdVxiLoTyK3a7cIkpuUFXNUrfldmzeo5UwsSvBQvnAVUgMRZplmtMQJuKwJbhmvsSU7HAQ8ZsGsFjuDa8pmPQtE4BZEJR48YeVHx3U9SQaq9GVz+zQKlpTNhI0lllp3+uDYdHo/F4dOZV1ipfqfNrfTxcq7JW+74W2cuc13in9yBq+Qoctfz4VHA7vrWT3YEiXG0olHRazmf0DtnIFuS9Kov7RGpUlPDtYLMDW9vtwh5lzO6sMqLVnnUSSi/ZKl7L3CszFJ6cDofXh+fH16NP4+Hp+cDLi1wzyvELTfR8D17/Ufi9hAIFZesEWFWY/3767YH4wpyveVtkvHLu/s+zbqvrH5f2MnEPy5hFx5VJib/nqPQl8uJku/eH28bGhutPVu3Pxk9+RZyoJY+hrHzW3KEOYiK1mEmSzWnsS50v5sIczNXpsDxO9ytiRD8Q8yeqD952dq4RdIfuI60XBL5rojMupGkbl+tNWztcLbwqGdB7kO/X9vy8dJ1AwATvGEc7sBlutldunVd6VO/as71crV+ruJOiRqkuULpFhxYW79aAMZopQZPwy8erdzthSu5omlv3clXB0i0+tMAb+P+61op0IWhY5hc4I3oexkIFjOh2+4HBehfVr6QqKFes41Pof/VBf4MekFtCdWHEjfv+7UwofYyaUIaue4CiCfQwXUixoAlKv1W+ulW/2Y/V5YqK8GxC/RZ1zq1o2Vkb9zxcG/b5/PDBRFv6YH6/pBsotqfzeE54wtb71KtYIvKrjMQ4WCDXvzqhwG33mPAFsXa8cqhQn/Is14euGQrsvmhD7311O0qyXDXXMUlRknCG+oLGN5dk6XTCorhWQY0LCMu+zDXjGY1vAkmWHTfqVOgUglc+jASnlGMSlAbavj4YoHQuudW4Nz+KYtJAMdfIFiaqxgySnccwGy8zDIeDk/F1f3ja/0fZ75uSmOCU5MxEZlCwmMBPlcQ6mgq1uRy5gypBpSkn7qRqJuMxziSiCja2trfCt7tvdzuw8yZ8t7u93YGdv73ZDN+WN6TSdcqppoStF0Lk2tYe+8l4ZS95T7m+Xsc/e6W+NRXUKrljXJl+97HKELd875GT06XcaYUxo8i1PeWgC9udhknXE5SzVS4+R2LryP8NeX1ev5i6mHNNWXmN15Shgtwc6NUEz8kCgQmSYOJvDeWp4oqStUcVpEg45bNpzjpGikMiRWbl6wRS5Z3V3Rdc9odUaeQ2o1XMjFNDQZILKWbmtmz3mWnN7B+Fkk138HuOuXlKKWqQA9eNQq/Xg80STmhc23c4tjlrZqiftym499fIcp9oIdiEyJ9ud8p9Ym7FhbEeJCLOUxP1DPWAofnzaHmaBFHLy0Qt44Ylkh/xcTbYiCUSjd5MELUSunDqpXRIE+hBxXht0naYZcm2gmSiBMs1NkpqkVmh3eyucZ7hVD8pMCHxjX9OMmJyNiHBjilW/nszfLfbbnaSJPZia9TePGZdyMS1JbmygjulYLHhi7emkGQZ8qQ/pywJCpzrB6t9GoQefDX63909wtwAXptLAaRCZvM9CCxBHcft0Fi8Pg5MD2MZVVPsm8Z7kiswPHzaRiFqT4Nma9vPuLFdcePbfj2sj1Lk2YuIVEqHMSNKnRP7rFE83W1Mcq0F3zApzSzSRUKq8JY22mtuHFltg7F7hk1JFljp9c5j8pSvzoeC95OnHPVkmITmpmfeHZFrv3pogfXTgrs7T291WQnKamOlLchlBakEE06FHJB4HgRCz1GaSIoqZAece6ZG+QeWYB1MEpsHq6hV3BSLRSYVVftI9pieO1fsz1X6qgmZOBF3CsFk37cIlSi+bn77gdUqbVHz66oWsxl7+c2x4EiiiXYOvTT/K41naVARddUjJXJG+diXuKLGlATIs4RoPNZE25t4wYeKmTqpfnsEjD348/eXvEh/gKg1Oo9aYDb76OQkat3/tr+24ipFDmHfKD2Wq+IW8vzqZdt0XwPqiX3xskf2Vy8Qs0E+QPvp14r9iDdpNFWjVTg16q6/nHrK+pbs41XXQfqn3c2d7TY8S92J+jHiFvLP0rYU/AHSHqkaZUsTa4Q9qkNgifrwNf9RXtZDqLHyMT42vI970pXGnqJc078aXjU+uT/6/rUG0AtZVpVv4ljhvZlvdVoHNlf2HPi7/+9ZLlkQhl2NacaIRtWd5PEN6jBW9l5+0C1UDhK6AJr0Gv5bBhboXtSa5oxd0X9h1Hp/0E3ooqZmmn3KZ6MFSkaWRmS+9X7oBsMwPOjOtxq0yn6xnPsP), which was AI generated.

Video Before and After

Shows the scenes working in 3D, then 2D, then the red wireframe shows the bounding sphere.

https://github.com/user-attachments/assets/552efcf3-85c0-4a4f-9721-9f6f2d5eb360

## Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

## AI acknowledgment

- [x] I used AI to generate content in this PR
- [x] If yes, I have reviewed the AI-generated content before submitting

If yes, I used the following Tools(s) and/or Service(s):

Crush was used to generate the sandcastle and unit tests.

If yes, I used the following Model(s):

Claude Opus 4.8

### CODE REVIEW

**Comment by chanwooro** [comment]
👍

**Comment by lukemckinstry** [comment]
This is looking good @harrythemorris. Fixes all the test cases in the linked issues. Still reviewing some code and checking for regressions, but this looks like a great fix!

File: packages/engine/Source/Scene/Billboard.js [hunk 1]
```diff
@@ -1375,7 +1372,17 @@ Billboard._computeActualPosition = function (
...
+    // Clamped position is already our rendering position when in 3D
+    if (frameState.mode === SceneMode.SCENE3D) {
+      return billboard._clampedPosition;
+    }
+
+    // in 2D and Columbus View we instead project the ECEF coordinate into the current map frame.
+    return SceneTransforms.computeActualEllipsoidPosition(
+      frameState,
+      billboard._clampedPosition,
+    );
```
Comment (mzschwartz5) [inline packages/engine/Source/Scene/Billboard.js:1385]: Suggestion: without passing in a scratch result, this will allocate a new `Cartesian3` per clamped billboard per frame. That'll lead to really bad GC pauses for moderately sized collections.

File: CHANGES.md [hunk 2]
```diff
@@ -20,6 +20,7 @@
 - Auto-normalize non-unit `alignedAxis` in `BillboardCollection` instead of silently ignoring it. [#6596](https://github.com/CesiumGS/cesium/issues/6596)
 - Fixed SPZ-compressed Gaussian splat loading to read the compressed payload from the buffer view declared by `KHR_gaussian_splatting_compression_spz_2`, preventing incorrect cache reuse for assets with SPZ payloads in different buffer views. [#12847](https://github.com/CesiumGS/cesium/issues/12847)
 - Fixed incorrect parameter order when calling `clamp` in `PolylineGlowMaterial` shader, which caused the alpha value to always be 1.0 regardless of the glow intensity.
+- Fixed terrain-clamped billboards, labels, and points being mispositioned and not properly rendering in 2D/Columbus view. [#5042](https://github.com/CesiumGS/cesium/issues/5042) [#12531](https://github.com/CesiumGS/cesium/issues/12531)
```
Comment (mzschwartz5) [inline CHANGES.md:23]: Does this PR affect points? I don't see any evidence that this is the case. (Points are a separate, though similar pipeline from billboards and labels)

File: packages/engine/Source/Shaders/BillboardCollectionVS.glsl [hunk 3]
```diff
@@ -292,7 +292,9 @@ void main()
     v_compressed.y = enableDepthCheck;
 
 #ifdef VS_THREE_POINT_DEPTH_CHECK
-if (lengthSq < (u_threePointDepthTestDistance * u_threePointDepthTestDistance) && (enableDepthCheck == 1.0)) {
+// In 2D, the globe and billboards are coplanar, but the depth reconstruction from getGlobeDepth is imprecise
+// It suffered error greater than the depthsilon of 10.0, causing the verts to be discarded for clamped billboards.
+if (czm_sceneMode != czm_sceneMode2D && lengthSq < (u_threePointDepthTestDistance * u_threePointDepthTestDistance) && (enableDepthCheck == 1.0)) {
```
Comment (mzschwartz5) [inline packages/engine/Source/Shaders/BillboardCollectionVS.glsl:297]: This conditional is _quite_ complex now. I'd recommend just creating a small, well-named function to contain it for readability's sack.

File: packages/engine/Source/Scene/QuadtreePrimitive.js [hunk 4]
```diff
@@ -1527,21 +1530,41 @@ function updateHeights(primitive, frameState) {
...
+                scratchCartographic,
+              );
+            }
+            if (defined(positionCarto)) {
+              tile.setPositionCacheEntry(
+                data.positionCartographic,
+                primitive.maximumScreenSpaceError,
+                Cartographic.clone(positionCarto),
+              );
+            }
```
Comment (mzschwartz5) [inline packages/engine/Source/Scene/QuadtreePrimitive.js:1563]: This is a pretty gnarly area of code already. In an effort to keep it maintainable, and to make this new logic reusable, I think a new `SceneTransforms` helper could be in order. E.g. this call site could become

```JS
positionCarto = SceneTransforms.actualEllipsoidPositionToCartographic(
  frameState,
  position,
  scratchCartographic,
);
```

Where `SceneTransforms.actualEllipsoidPositionToCartographic` is something like:


```JS
SceneTransforms.actualEllipsoidPositionToCartographic = function (
  frameState,
  position,
  result,
) {
  const projection = frameState.mapProjection;

  if (frameState.mode === SceneMode.SCENE3D) {
    return projection.ellipsoid.cartesianToCartographic(position, result);
  }

  // Un-swizzle the (height, easting, northing) render frame back to the projection's
  // native (easting, northing, height) layout before unprojecting.
  const projected = Cartesian3.fromElements(
    position.y,
    position.z,
    position.x,
    inverseProjectedScratch,
  );

  return projection.unproject(projected, result);
};
```

**Review by mzschwartz5 — CHANGES_REQUESTED** [review]
Hi @harrythemorris -- thanks for the contribution!

On the whole, looks good. (Though, I have to admit: the areas of code you've modified are already quite complex, so it's very difficult to give a thorough review).

I left a few suggestions that I think will simplify and improve your changes. ~Also, one thing I noticed while testing: it looks like this change doesn't fix CV mode (or at least, not all the time?). E.g:~

EDIT: I'm not able to reproduce the CV issue anymore, not sure what I'm doing differently but it seems fine. Maybe I was testing on main before...?

Image

~This is still a good improvement. If fixing CV isn't straightforward, let's just be upfront about this being for 2D in the change log.~

**Comment by harrythemorris** [comment]
Thanks for the reviews @lukemckinstry @mzschwartz5!!

I'll take a pass and see if I can improve readability - was a bit of a toss up between minimal diff or big scary refactor to make it understandable 😅

**Comment by lukemckinstry** [comment]
> It looks like this change doesn't fix CV mode (or at least, not all the time?).

@mzschwartz5 do you have a sandcastle link for the example you shared the screenshot from? CV was looking good in the linked issues and other examples I could find to test, but if there are edge cases it would be good to have reproducible links.

**Comment by mzschwartz5** [comment]
@lukemckinstry The screenshot came from the sandcastle linked from the testing instructions

**Comment by mzschwartz5** [comment]
Hi @harrythemorris -- just wanted to see if you're planning to make more changes here in response to the review. It may still be possible for these changes to go into the September release.

**Comment by lukemckinstry** [comment]
One issue I am seeing is inconsistent bounding volume calculation when switching between 3D/CV/2D modes

When I do various combinations of scene mode switching and camera movements + zooms, sometimes the bounding sphere appears incorrect as shown at the end of this clip. I don't see a consistent pattern for this.

Update: I do not think the bounding volume issue is specific to this PR and I do not see any noticeable impact on bahavior.

https://github.com/user-attachments/assets/7feb93bf-88aa-4e07-b4f5-a0501b2df732

**Comment by mzschwartz5** [comment]
@lukemckinstry I addressed my own feedback - would you mind taking another review pass?

**Review by lukemckinstry — APPROVED** [review]
Fixes all the linked test cases, no regressions found :rocket: 

Thanks for figuring all this out @harrythemorris and thanks @mzschwartz5  for the review and getting it over the finish line :pray:

**Review by mzschwartz5 — APPROVED** [review]
